# MODUL PRAKTIKUM 3: PENGOLAHAN DATA DENGAN NUMPY DAN PANDAS
**Mata Kuliah:** Visualisasi Data  
**Pertemuan Ke-:** 3 (tiga)  
**Topik:** Pengolahan Data dengan NumPy dan Pandas  
**Dataset:** Titanic Dataset (seaborn-data)  

### Identitas Mahasiswa
- **Nama Mahasiswa:** Gathan Hilabi
- **NIM:** 60324059
- **Kelas / Prodi:** S1 Informatika
- **Tanggal Praktikum:** 9 Oktober 2026

---


### 1. Menyiapkan Notebook dan Mengimpor Library
Mengimpor pustaka `numpy` dan `pandas` untuk analisis dan manipulasi data tabular.


In [15]:
import numpy as np
import pandas as pd


### 2. Dasar Array NumPy
Membuat array numerik dan menghitung ringkasan statistik dasar (rata-rata, maksimum, minimum, deviasi standar).


In [16]:
# Membuat array NumPy dari sebuah list
nilai_ujian = np.array([78, 85, 90, 65, 72])

print("Rata-rata:", np.mean(nilai_ujian))
print("Maksimum:", np.max(nilai_ujian))
print("Minimum:", np.min(nilai_ujian))
print(f"Std Deviasi: {np.std(nilai_ujian):.2f}")


Rata-rata: 78.0
Maksimum: 90
Minimum: 65
Std Deviasi: 8.92


### 3. Membaca Dataset dengan Pandas
Memuat Titanic dataset langsung dari URL publik GitHub (seaborn-data) dan menampilkan 5 baris pertama.


In [17]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.head() # menampilkan 5 baris pertama


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


### 4. Eksplorasi Awal Dataset
Mengetahui dimensi dataset, daftar kolom, tipe data masing-masing kolom, dan statistik deskriptif numerik.


In [18]:
print(df.shape) # jumlah baris & kolom
print(df.columns) # nama-nama kolom
df.info() # tipe data & jumlah data tidak kosong
df.describe() # statistik deskriptif kolom numerik


(891, 15)
Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 no

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### 5. Mengakses Kolom dan Baris
Mengambil kolom spesifik, memilih beberapa atribut sekaligus, dan melakukan filtering baris dengan boolean indexing.


In [19]:
# Mengakses satu kolom
print("Contoh kolom age:")
print(df["age"].head())

# Mengakses beberapa kolom sekaligus
print("\nContoh subset kolom age, sex, survived:")
print(df[["age", "sex", "survived"]].head())

# Memfilter baris berdasarkan kondisi (usia < 18)
penumpang_anak = df[df["age"] < 18]
print("\nJumlah penumpang anak:", len(penumpang_anak))


Contoh kolom age:
0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: age, dtype: float64

Contoh subset kolom age, sex, survived:
    age     sex  survived
0  22.0    male         0
1  38.0  female         1
2  26.0  female         1
3  35.0  female         1
4  35.0    male         0

Jumlah penumpang anak: 113


### 6. Mendeteksi Missing Value
Mengecek jumlah nilai kosong (null / NaN) pada masing-masing kolom dataset.


In [20]:
print("Jumlah missing value per kolom:")
print(df.isnull().sum())


Jumlah missing value per kolom:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


### 7. Menangani Missing Value
Menerapkan dua strategi pembersihan missing value: imputasi rata-rata pada kolom `age` dan penghapusan baris pada kolom `embarked`.


In [21]:
# Strategi 1: isi nilai kosong dengan rata-rata (untuk data numerik age)
df["age"] = df["age"].fillna(df["age"].mean())

# Strategi 2: hapus baris yang datanya kosong pada kolom tertentu (embarked)
df = df.dropna(subset=["embarked"])

print("Jumlah missing value setelah penanganan:")
print(df.isnull().sum()) # periksa kembali


Jumlah missing value setelah penanganan:
survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
embarked         0
class            0
who              0
adult_male       0
deck           688
embark_town      0
alive            0
alone            0
dtype: int64


### 8. Mendeteksi dan Menghapus Duplikasi
Mendeteksi apakah terdapat baris duplikat, lalu menghapusnya dengan `drop_duplicates()`.


In [22]:
print("Jumlah data sebelum:", len(df))
print("Jumlah duplikasi:", df.duplicated().sum())

df = df.drop_duplicates()
print("Jumlah data sesudah:", len(df))


Jumlah data sebelum: 889
Jumlah duplikasi: 107
Jumlah data sesudah: 782


### 9. Menyimpan Dataset yang Sudah Dibersihkan
Menyimpan DataFrame hasil pembersihan ke dalam file `titanic_bersih.csv`.


In [23]:
df.to_csv("titanic_bersih.csv", index=False)
print("Dataset bersih berhasil disimpan ke titanic_bersih.csv dengan total baris:", len(df))


Dataset bersih berhasil disimpan ke titanic_bersih.csv dengan total baris: 782


### 10. Mendokumentasikan Proses dan Ringkasan Pembersihan Data
## Ringkasan Pembersihan Data
- **Dataset awal** berisi **891** baris dan **15** kolom.
- **Kolom age** memiliki **177** data kosong, telah diisi dengan nilai rata-rata (**29.70** tahun).
- **Baris dengan kolom embarked kosong** berjumlah **2** baris dan telah dihapus.
- **Ditemukan 107 baris duplikat** dan telah berhasil dihapus.
- **Dataset akhir** berisi **782** baris dan siap digunakan untuk analisis serta visualisasi data selanjutnya.

---


## F. Latihan Mandiri
Bagian ini memuat penyelesaian 5 butir latihan mandiri sesuai modul praktikum.


#### Latihan 1
Tampilkan 10 baris pertama dan 5 baris terakhir dataset menggunakan `head(10)` dan `tail(5)`.


In [24]:
# Menampilkan 10 baris pertama dataset
df.head(10)


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.000000,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.000000,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.000000,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.000000,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.000000,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
5,0,3,male,29.699118,0,0,8.4583,Q,Third,man,True,NaN,Queenstown,no,True
6,0,1,male,54.000000,0,0,51.8625,S,First,man,True,E,Southampton,no,True
7,0,3,male,2.000000,3,1,21.0750,S,Third,child,False,NaN,Southampton,no,False
8,1,3,female,27.000000,0,2,11.1333,S,Third,woman,False,NaN,Southampton,yes,False
9,1,2,female,14.000000,1,0,30.0708,C,Second,child,False,NaN,Cherbourg,yes,False


In [25]:
# Menampilkan 5 baris terakhir dataset
df.tail(5)


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
885,0,3,female,39.000000,0,5,29.125,Q,Third,woman,False,NaN,Queenstown,no,False
887,1,1,female,19.000000,0,0,30.000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,29.699118,1,2,23.450,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.000000,0,0,30.000,C,First,man,True,C,Cherbourg,yes,True
890,0,3,male,32.000000,0,0,7.750,Q,Third,man,True,NaN,Queenstown,no,True


#### Latihan 2
Hitung rata-rata usia penumpang menggunakan NumPy (`np.mean(df['age'])`) dan Pandas (`df['age'].mean()`). Bandingkan apakah hasilnya sama.


In [26]:
mean_numpy = np.mean(df["age"])
mean_pandas = df["age"].mean()

print("Rata-rata usia (NumPy) :", mean_numpy)
print("Rata-rata usia (Pandas):", mean_pandas)
print("Apakah hasilnya sama? :", np.isclose(mean_numpy, mean_pandas))
print("Kesimpulan: Kedua metode menghasilkan nilai rata-rata yang identik.")


Rata-rata usia (NumPy) : 29.79479088310516
Rata-rata usia (Pandas): 29.79479088310516
Apakah hasilnya sama? : True
Kesimpulan: Kedua metode menghasilkan nilai rata-rata yang identik.


#### Latihan 3
Tampilkan data penumpang yang berada di kelas 1 (`pclass == 1`) dan selamat (`survived == 1`) menggunakan filter kondisi ganda (operator `&`).


In [27]:
# Filter kondisi ganda: pclass == 1 dan survived == 1
penumpang_kelas1_selamat = df[(df["pclass"] == 1) & (df["survived"] == 1)]

print(f"Jumlah penumpang Kelas 1 yang selamat: {len(penumpang_kelas1_selamat)} orang")
penumpang_kelas1_selamat.head()


Jumlah penumpang Kelas 1 yang selamat: 133 orang


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
1,1,1,female,38.000000,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
3,1,1,female,35.000000,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
11,1,1,female,58.000000,0,0,26.5500,S,First,woman,False,C,Southampton,yes,True
23,1,1,male,28.000000,0,0,35.5000,S,First,man,True,A,Southampton,yes,True
31,1,1,female,29.699118,1,0,146.5208,C,First,woman,False,B,Cherbourg,yes,False


#### Latihan 4
Isi missing value pada kolom `embarked` dengan nilai yang paling sering muncul (modus), menggunakan `df['embarked'].mode()[0]`.


In [28]:
# Menunjukkan teknik pengisian modus pada kolom kategorikal (embarked)
# Menggunakan dataset contoh untuk mendemonstrasikan penanganan missing value secara eksplisit
df_latihan = pd.read_csv(url)
modus_embarked = df_latihan["embarked"].mode()[0]
print(f"Modus kolom embarked: '{modus_embarked}'")
print("Missing value sebelum imputasi:", df_latihan["embarked"].isnull().sum())

# Mengisi missing value dengan nilai modus
df_latihan["embarked"] = df_latihan["embarked"].fillna(modus_embarked)
print("Missing value setelah imputasi:", df_latihan["embarked"].isnull().sum())

# Terapkan juga pada dataframe utama jika masih ada nilai kosong
df["embarked"] = df["embarked"].fillna(df["embarked"].mode()[0])


Modus kolom embarked: 'S'
Missing value sebelum imputasi: 2
Missing value setelah imputasi: 0


#### Latihan 5
Hitung berapa persentase penumpang yang selamat dari keseluruhan data, menggunakan `df['survived'].mean() * 100`.


In [29]:
persentase_selamat = df["survived"].mean() * 100
total_selamat = df["survived"].sum()
total_penumpang = len(df)

print(f"Total penumpang : {total_penumpang}")
print(f"Penumpang selamat : {total_selamat}")
print(f"Persentase penumpang selamat: {persentase_selamat:.2f}%")


Total penumpang : 782
Penumpang selamat : 321
Persentase penumpang selamat: 41.05%


---
## H. Pertanyaan Diskusi

### 1. Mengapa penting memeriksa dan menangani missing value sebelum data digunakan untuk membuat visualisasi?
**Jawaban:**
Memeriksa dan menangani *missing value* sangat penting karena:
1. **Mencegah Error dan Inkonsistensi Visualisasi:** Banyak pustaka visualisasi (seperti Matplotlib, Seaborn, Plotly) atau model machine learning dapat mengalami *error*, melewatkan baris tertentu secara diam-diam, atau menghasilkan agregasi statistik yang terdistorsi saat menemui nilai `NaN`.
2. **Menjaga Akurasi dan Validitas Representasi Data:** Data yang hilang secara sistematis (misalnya kelompok demografi tertentu tidak mengisi form) dapat menimbulkan bias interpretasi jika tidak ditangani dengan sengaja.
3. **Transparansi Integritas Data:** Mengetahui persentase missing value membantu analis memutuskan apakah variabel tersebut masih layak dipertahankan atau sebaiknya dieliminasi.

### 2. Dalam kondisi seperti apa sebaiknya kita menggunakan `dropna()`, dan dalam kondisi seperti apa lebih baik menggunakan `fillna()`?
**Jawaban:**
- **Gunakan `dropna()` ketika:**
  - Proporsi baris yang memiliki nilai kosong sangat kecil (misalnya < 1-2% dari total dataset), sehingga penghapusannya tidak akan mengubah distribusi data secara signifikan (contoh: 2 baris pada kolom `embarked`).
  - Kolom tersebut merupakan variabel target/kunci yang tidak boleh diasumsikan atau ditebak.
  - Kolom memiliki persentase missing value yang terlampau tinggi (misalnya > 70-80% seperti kolom `deck` pada Titanic), sehingga kolom tersebut lebih baik di-drop sepenuhnya.
- **Gunakan `fillna()` ketika:**
  - Ukuran dataset relatif terbatas sehingga membuang baris akan mengurangi ukuran sampel secara drastis.
  - Missing value dapat diestimasi secara logis menggunakan parameter statistik tanpa merusak distribusi data, seperti menggunakan **mean** atau **median** untuk fitur numerik yang mendekati normal/miring, atau **modus** untuk fitur kategorikal.

### 3. Apa risiko yang bisa terjadi jika data duplikat tidak dihapus sebelum data dianalisis atau divisualisasikan?
**Jawaban:**
Jika data duplikat tidak dihapus, dampak negatif yang dapat terjadi meliputi:
1. **Distorsi Statistik Deskriptif:** Nilai rata-rata, deviasi standar, kuartil, dan total agregasi akan terboboti secara berlebihan ke observasi yang terduplikasi (*pseudoreplication*).
2. **Bias Visualisasi:** Plot seperti histogram, boxplot, heatmap korelasi, dan countplot akan menampilkan densitas atau frekuensi palsu yang tidak mencerminkan populasi sebenarnya.
3. **Overfitting pada Pemodelan:** Jika data duplikat terbawa ke pemodelan prediktif, model dapat menghafal data duplikat tersebut sehingga terjadi kebocoran data (*data leakage*) antara data latih dan data uji.
